### PR TITLE
Merge small islands with colliding continents

### PR DESCRIPTION
**Pull request by pjanik**
Demo:
http://models-resources.concord.org/plate-tectonics-3d/branch/islands/index.html?preset=islandCollision
You can notice that islands are merged with the colliding continent. Before they were ignored and just subducted under the plate.

Important change is that now `.island` property is calculated dynamically in each model step. It ensures it'd work both for islands that are created due to volcanic activity and islands that are defined using image or continents drawing (although current tool doesn't let you draw so small areas).

### CODE REVIEW

File: js/plates-model/plate.js [hunk 1]
```diff
@@ -179,6 +181,75 @@ export default class Plate extends PlateBase {
...
+
+  mergeIsland (island, collidingField) {
+    const perfectPosition = island.absolutePos
+    let bestFieldId = null
+    let minDist = Infinity
+
+    for (let adjId of collidingField.adjacentFields) {
+      const adjField = this.adjacentFields.get(adjId)
+      if (adjField) {
+        if (adjField.absolutePos.distanceTo(perfectPosition) < minDist && adjField.neighboursCount() > 1) {
```
Comment (ekosmin) [inline js/plates-model/plate.js]: What's the purpose of the `adjField.neighboursCount() > 1` check? To make sure you don't collect islands in some kind of spike?

File: js/plates-model/plate.js [hunk 2]
```diff
@@ -179,6 +181,75 @@ export default class Plate extends PlateBase {
     return count
   }
 
+  markIslands () {
+    // DFS-based algorithm which calculates area of continents and mark small ones as islands.
```
[the remark this answers is not in the document]
Comment (pjanik) [reply js/plates-model/plate.js]: Right, just did it.

File: js/plates-model/plate.js [hunk 3]
```diff
@@ -179,6 +181,75 @@ export default class Plate extends PlateBase {
...
+        visited[field.id] = true
+        continentId[field.id] = field.id
+        area[continentId[field.id]] = 0
+        calcAreaOfContinent()
+      }
+    })
+
+    this.forEachField(field => {
+      const cId = continentId[field.id]
+      if (cId !== undefined && area[cId] < MAX_ISLAND_SIZE) {
```
[the remark this answers is not in the document]
Comment (pjanik) [reply js/plates-model/plate.js]: Yes, I've added a comment so it's clear.

File: js/plates-model/plate.js (hunk 1 above)
Comment (pjanik) [reply js/plates-model/plate.js]: Exactly. Narrow spikes don't look very realistic and it's more likely that we end up with a plate with some holes. I've added a comment to clarify that too.

**Pull request merged**